sdk/identity/azure_identity/src/client_secret_credential.rs

```diff
@@ -0,0 +1,276 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+use crate::{credentials::cache::TokenCache, EntraIdTokenResponse};
+use crate::{EntraIdErrorResponse, TokenCredentialOptions};
+use azure_core::http::{Response, StatusCode};
+use azure_core::Result;
+use azure_core::{
+    credentials::{AccessToken, Secret, TokenCredential},
+    error::{ErrorKind, ResultExt},
+    http::{
+        headers::{self, content_type},
+        Method, Request, Url,
+    },
+    Error,
+};
+use std::time::Duration;
+use std::{str, sync::Arc};
+use time::OffsetDateTime;
+use url::form_urlencoded;
+
+/// Options for constructing a new [`ClientSecretCredential`].
+#[derive(Debug, Default)]
+pub struct ClientSecretCredentialOptions {
+    /// Options for constructing credentials.
+    pub credential_options: TokenCredentialOptions,
+}
+
+/// Authenticates an application with a client secret.
+#[derive(Debug)]
+pub struct ClientSecretCredential {
+    cache: TokenCache,
+    client_id: String,
+    endpoint: Url,
+    options: TokenCredentialOptions,
+    secret: Secret,
+}
+
+impl ClientSecretCredential {
+    pub fn new(
+        tenant_id: String,
+        client_id: String,
```

Comment:
`tenant_id`, at least, should be a `&str` because we don't take ownership of it. We probably should declare `secret` as either `Secret` - make the caller do the allocation as needed, if even needed (probably) - or this may be a case to declare it as `impl Into<Secret>` which allows for a lot of possibilities. I'm leaning toward the former because it better follows our guidelines.

Reply:
QQ: (since this hits some of the stuff I'm doing): Should we be consistent across all parameters to the function? To me, it feels weird to have `foo(a:String, b:&str, c:String)` just because of what the function does with the different parameters (especially since the caller's experience is different for each - it feels like you're exposing an implementation detail of the API - and if at some point in the future you need to take ownership of `tenant_id`, you're a bit stuck).

Reply:
This is a separate thread I started but for now I would say: follow guidelines. If we're going to own it, make them pass a concrete struct; otherwise, pass a ref.

Reply:
I agree with Larry on his concern about exposing the implementation detail. We are forever bound to the api design. Is it really worth making the customer think about how to pass a string because we may/may not want ownership?